src/jit/morph.cpp

```diff
@@ -4440,113 +4690,134 @@ GenTreePtr    Compiler::fgMorphMultiregStructArg(GenTreePtr arg, fgArgTabEntryPt
-                //
-                newArg = gtNewListNode(loLclFld, gtNewArgList(hiLclFld));
-            }
         }
     }
-    // Check if we already created a replacement newArg above
+
+    // If we didn't set newarg to a new List Node tree
+    //
     if (newArg == nullptr)
     {
         if (fgEntryPtr->regNum == REG_STK)
         {
             // We leave this stack passed argument alone
             return arg;
         }
-    }
 
-    // Are we passing a GT_LCL_FLD which contain a 16-byte struct inside it?
-    //
-    if (argValue->OperGet() == GT_LCL_FLD)
-    {
-        GenTreeLclVarCommon* varNode = argValue->AsLclVarCommon();
-        unsigned   varNum = varNode->gtLclNum;
-        assert(varNum < lvaCount);
-        LclVarDsc* varDsc = &lvaTable[varNum];
-
-        unsigned baseOffset   = argValue->gtLclFld.gtLclOffs;
-        unsigned baseIndex    = baseOffset / TARGET_POINTER_SIZE;
-        unsigned requiredSize = baseOffset + (2 * TARGET_POINTER_SIZE);
+        // Are we passing a GT_LCL_FLD (or a GT_LCL_VAR that was not struct promoted )
+        // A GT_LCL_FLD could also contain a 16-byte struct or HFA struct inside it?
+        //
+        if ((argValue->OperGet() == GT_LCL_FLD) || (argValue->OperGet() == GT_LCL_VAR))
+        {
+            GenTreeLclVarCommon* varNode = argValue->AsLclVarCommon();
+            unsigned   varNum = varNode->gtLclNum;
+            assert(varNum < lvaCount);
+            LclVarDsc* varDsc = &lvaTable[varNum];
 
-        // The allocated size of our LocalVar must be at least as big as requiredSize
-        assert(varDsc->lvSize() >= requiredSize);
+            unsigned baseOffset = (argValue->OperGet() == GT_LCL_FLD) ? argValue->gtLclFld.gtLclOffs : 0;
+            unsigned lastOffset = baseOffset + (elemCount * elemSize);
 
-        const BYTE * gcPtrs = varDsc->lvGcLayout;
+            // The allocated size of our LocalVar must be at least as big as lastOffset
+            assert(varDsc->lvSize() >= lastOffset);
 
-        var_types type0 = getJitGCType(gcPtrs[baseIndex+0]);
-        var_types type1 = getJitGCType(gcPtrs[baseIndex+1]);
+            if (varDsc->lvStructGcCount > 0)
+            {
+                // alignment of the baseOffset is required
+                noway_assert((baseOffset % TARGET_POINTER_SIZE) == 0);
+                noway_assert(elemSize == TARGET_POINTER_SIZE);
+                unsigned baseIndex = baseOffset / TARGET_POINTER_SIZE;
+                const BYTE * gcPtrs = varDsc->lvGcLayout;  // Get the GC layout for the local variable
+                for (unsigned inx = 0; (inx < elemCount); inx++)
+                {
+                    // The GC information must match what we setup using 'objClass'
+                    noway_assert(type[inx] == getJitGCType(gcPtrs[baseIndex + inx]));
+                }
+            }
+            else //  this varDsc contains no GC pointers
+            {
+                for (unsigned inx = 0; (inx < elemCount); inx++)
```

Comment:
(Minor) to be consistent, i would suggest to remove braces around for-loop condition above and also on line 4735